src/rust/engine/graph/src/lib.rs

```diff
@@ -75,77 +76,141 @@ pub struct Entry<N: Node> {
-          let context = context.clone_for(entry_id);
-          let node = n.clone();
-          future::lazy(move || node.run(context)).to_boxed()
+    let next_state = match &mut self.state {
+      &mut EntryState::NotStarted(run_token) => {
+        match &self.node {
+          &EntryKey::Valid(ref n) => {
+            let context2 = context.clone_for(entry_id);
+            let node = n.clone();
+
+            // Spawn the execution of the work on an Executor, which will cause it to execute
+            // outside of the Graph lock, and allow us to call back into the graph lock to set the
+            // final value.
+            context.spawn(future::lazy(move || {
+              let context = context2.clone();
+              node.run(context2).then(move |res| {
+                context.graph().complete(entry_id, run_token, res);
+                Ok(())
+              })
+            }));
+
+            EntryState::Running {
+              waiters: Vec::new(),
+              start_time: Instant::now(),
+              run_token,
+            }
+          }
+          &EntryKey::Cyclic(_) => EntryState::Completed {
+            result: Err(N::Error::cyclic()),
+            run_token: run_token,
+          },
         }
-        &EntryKey::Cyclic(_) => future::err(N::Error::cyclic()).to_boxed(),
-      };
+      }
+      &mut EntryState::Running {
+        ref mut waiters, ..
+      } => {
+        let (send, recv) = oneshot::channel();
+        waiters.push(send);
+        return recv
+          .map_err(|_| N::Error::invalidated())
+          .and_then(|res| res)
+          .to_boxed();
+      }
+      &mut EntryState::Completed { ref result, .. } => {
+        return future::result(result.clone()).to_boxed();
+      }
+    };
 
-      self.state = Some(EntryState {
-        field: state.shared(),
-        start_time,
-      });
-      self.state(context, entry_id)
-    }
+    // swap in the new state and then recurse.
+    self.state = next_state;
+    self.get(context, entry_id)
   }
 
   ///
   /// If the Future for this Node has already completed, returns a clone of its result.
   ///
   fn peek(&self) -> Option<Result<N::Item, N::Error>> {
-    self
-      .state
-      .as_ref()
-      .and_then(|state| state.field.peek().map(unwrap_entry_res::<N>))
+    match &self.state {
+      &EntryState::Completed { ref result, .. } => Some(result.clone()),
+      _ => None,
+    }
+  }
+
+  ///
+  /// Called from the Executor when a Node completes.
+  ///
+  fn complete(&mut self, result_run_token: RunToken, result: Result<N::Item, N::Error>) {
+    // Temporarily swap in `NotStarted` in order to take ownership of the state.
+    let state = mem::replace(&mut self.state, EntryState::NotStarted(0));
+
+    // We care about exactly one case: a Running state with the same run_token. All other states
+    // represent various (legal) race conditions.
```

Comment:
Can you expand on those race conditions? (i.e. mention that it means a thing was invalidated and we don't care about its result any more)